File: benchmark.py
import os
import subprocess
import time
import matplotlib.pyplot as plt
import re
import pandas as pd # Import pandas

# --- Configuration ---
C_FILES = [
    "bubble_sort.c",
    "heap_sort.c",
    "insertion_sort.c",
    "merge_sort.c",
    "quick_sort.c", # This is the generic quick sort (last element pivot)
    "radix_sort.c",
    "selection_sort.c",
]
C_FILE_PATHS = [os.path.join("sorting_algorithms", f) for f in C_FILES]
EXECUTABLES_DIR = "executables"
TEST_DATA_DIR = "test_data"
OUTPUT_GRAPHS_DIR = "graphs"
NUM_REPETITIONS = 7  # Number of times to run each experiment for averaging
# The maximum input size to consider for plotting. Adjust if needed.
MAX_PLOT_N = 100000 

# --- Helper Functions ---

def compile_c_code(c_file_path, output_executable_path):
    """Compiles a C source file into an executable."""
    print(f"Compiling {c_file_path}...")
    try:
        subprocess.run(
            ["gcc", c_file_path, "-o", output_executable_path],
            check=True,
            capture_output=True,
            text=True
        )
        print(f"Successfully compiled {c_file_path} to {output_executable_path}")
    except subprocess.CalledProcessError as e:
        print(f"Error compiling {c_file_path}:")
        print(f"Stdout: {e.stdout}")
        print(f"Stderr: {e.stderr}")
        exit(1)

def read_test_data(filepath):
    """Reads integers from a given test data file, one integer per line."""
    with open(filepath, 'r') as f:
        data = [int(line.strip()) for line in f if line.strip()] # Read all non-empty lines as integers
    return data

def run_benchmark(executable_path, data):
    """
    Runs the compiled C program with the given data, measures execution time.
    Input data is passed via stdin. Output is discarded.
    """
    input_str = f"{len(data)}\n" + " ".join(map(str, data))
    
    start_time = time.perf_counter()
    try:
        # Use Popen to send input via stdin and capture/discard stdout/stderr
        process = subprocess.run(
            [executable_path],
            input=input_str,
            capture_output=True, # Capture stdout and stderr
            text=True,
            check=True # Raise an exception for non-zero exit codes
        )
        # We don't care about the sorted output for timing, just that it ran successfully
    except subprocess.CalledProcessError as e:
        print(f"Error running {executable_path} with data size {len(data)}:")
        print(f"Stdout: {e.stdout}")
        print(f"Stderr: {e.stderr}")
        return float('inf') # Return infinity for failed runs
    end_time = time.perf_counter()
    return end_time - start_time

def extract_n_and_type(filename):
    """Extracts N and data type (random, sorted, reverse_sorted) from a filename."""
    match = re.match(r"n_(\d+)_(\w+)\.txt", filename) # Corrected regex: \.txt instead of \\.txt
    if match:
        return int(match.group(1)), match.group(2)
    return None, None

def plot_results(results, plot_type, output_dir):
    """Generates and saves a plot for a specific case (best, worst, average)."""
    plt.figure(figsize=(12, 7))
    
    for algo_name, data_points in results.items():
        # Sort data points by N for correct plotting
        data_points.sort(key=lambda x: x[0])
        ns = [dp[0] for dp in data_points if dp[0] <= MAX_PLOT_N]
        times = [dp[1] for dp in data_points if dp[0] <= MAX_PLOT_N]
        plt.plot(ns, times, marker='o', linestyle='-', label=algo_name)

    plt.xlabel("Input Size (n)")
    plt.ylabel("Average Execution Time (s)")
    plt.title(f"Sorting Algorithm Performance: {plot_type} Case")
    plt.legend()
    plt.grid(True)
    plt.xscale('log') # Use log scale for x-axis if N values span a wide range
    plt.tight_layout()
    
    output_path = os.path.join(output_dir, f"sorting_algorithms_{plot_type.lower().replace(' ', '_')}_case.png")
    plt.savefig(output_path)
    print(f"Generated plot: {output_path}")
    plt.close()

# --- Main Execution ---
if __name__ == "__main__":
    # Create necessary directories
    os.makedirs(EXECUTABLES_DIR, exist_ok=True)
    os.makedirs(OUTPUT_GRAPHS_DIR, exist_ok=True)

    executables = {}
    for c_file in C_FILES:
        base_name = os.path.splitext(c_file)[0]
        executable_path = os.path.join(EXECUTABLES_DIR, base_name)
        compile_c_code(os.path.join("sorting_algorithms", c_file), executable_path)
        executables[base_name] = executable_path

    all_results = {
        "bubble_sort": {"random": [], "sorted": [], "reverse_sorted": []},
        "heap_sort": {"random": [], "sorted": [], "reverse_sorted": []},
        "insertion_sort": {"random": [], "sorted": [], "reverse_sorted": []},
        "merge_sort": {"random": [], "sorted": [], "reverse_sorted": []},
        "quick_sort": {"random": [], "sorted": [], "reverse_sorted": []},
        "radix_sort": {"random": [], "sorted": [], "reverse_sorted": []},
        "selection_sort": {"random": [], "sorted": [], "reverse_sorted": []},
    }

    # Get all test data files
    test_data_files = [f for f in os.listdir(TEST_DATA_DIR) if f.endswith(".txt")]
    test_data_files.sort(key=lambda x: (extract_n_and_type(x)[0], extract_n_and_type(x)[1]))

    print("\nStarting benchmarking...")
    for data_file in test_data_files:
        n, data_type = extract_n_and_type(data_file)
        if n is None:
            continue
        
        data_filepath = os.path.join(TEST_DATA_DIR, data_file)
        original_data = read_test_data(data_filepath)
        
        print(f"Benchmarking N={n}, Type={data_type}...")

        for algo_base_name, executable_path in executables.items():
            total_time = 0
            for _ in range(NUM_REPETITIONS):
                # Pass a copy of the data because C program might modify it
                time_taken = run_benchmark(executable_path, list(original_data))
                total_time += time_taken
            
            avg_time = total_time / NUM_REPETITIONS
            print(f"  {algo_base_name}: Average time = {avg_time:.6f} s")
            
            # Store results
            all_results[algo_base_name][data_type].append((n, avg_time))

    print("\nBenchmarking complete. Generating plots...")

    # --- Plotting ---
    # Define which data types correspond to best, worst, and average cases for plotting
    # This is a simplification and depends on the specific algorithm.
    # For general sorting algorithms:
    #   - Average Case: Random data
    #   - Worst Case: Reverse-sorted data (for many comparison sorts)
    #   - Best Case: Sorted data (for many comparison sorts)

    # Plot Average Case (using random data)
    avg_case_plot_data = {}
    for algo, types in all_results.items():
        avg_case_plot_data[algo.replace('_', ' ').title()] = types["random"]
    plot_results(avg_case_plot_data, "Average Case (Random Input)", OUTPUT_GRAPHS_DIR)

    # Plot Worst Case (using reverse_sorted data)
    worst_case_plot_data = {}
    for algo, types in all_results.items():
        worst_case_plot_data[algo.replace('_', ' ').title()] = types["reverse_sorted"]
    plot_results(worst_case_plot_data, "Worst Case (Reverse Sorted Input)", OUTPUT_GRAPHS_DIR)

    # Plot Best Case (using sorted data)
    best_case_plot_data = {}
    for algo, types in all_results.items():
        best_case_plot_data[algo.replace('_', ' ').title()] = types["sorted"]
    plot_results(best_case_plot_data, "Best Case (Sorted Input)", OUTPUT_GRAPHS_DIR)

    print("\nAll plots generated successfully.")
    print(f"Results are in the '{OUTPUT_GRAPHS_DIR}' directory.")
    print("\n--- Benchmarking Methodology ---")
    print(f"Each experiment was repeated {NUM_REPETITIONS} times, and the average execution time is reported.")
    print("Timing mechanism: Python's `time.perf_counter()` for high-resolution timing.")
    print("Input selection: Pre-generated test data from the 'test_data/' directory was used.")
    print("Same inputs were used for all sorting algorithms to ensure a fair comparison.")
    print("\nNote on Best/Worst Case Definitions:")
    print("  - Average Case: Represented by 'random' input data.")
    print("  - Worst Case: Represented by 'reverse_sorted' input data. For many comparison sorts (e.g., Bubble, Insertion, Selection, Quick Sort with last/first pivot), this is a true worst-case scenario.")
    print("  - Best Case: Represented by 'sorted' input data. For many comparison sorts (e.g., Insertion, Merge, Heap), sorted data often leads to good performance. However, for Quick Sort with a naive pivot (like the 'quick_sort.c' which uses the last element), sorted data can be a WORST-CASE scenario. The plots will reflect this behavior.")

    # --- Generate Table ---
    print("\n--- Benchmarking Results Table ---")
    table_data = []
    for algo_key, types_data in all_results.items():
        algo_name = algo_key.replace('_', ' ').title()
        for data_type, results_list in types_data.items():
            for n, avg_time in results_list:
                table_data.append({
                    "Algorithm": algo_name,
                    "Input Type": data_type.replace('_', ' ').title(),
                    "Input Size (N)": n,
                    "Average Time (s)": f"{avg_time:.6f}" if avg_time != float('inf') else "Crashed/Timeout"
                })
    
    df = pd.DataFrame(table_data)
    df_sorted = df.sort_values(by=["Input Size (N)", "Input Type", "Algorithm"])
    print(df_sorted.to_string(index=False))

    # Optionally save to CSV
    csv_output_path = os.path.join(OUTPUT_GRAPHS_DIR, "all_sorting_benchmark_results.csv")
    df_sorted.to_csv(csv_output_path, index=False)
    print(f"\nFull results table saved to {csv_output_path}")

File: benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <stdbool.h>

// --- Helper Functions ---

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

bool is_sorted(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) return false;
    }
    return true;
}

double get_time_seconds() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

// --- Bubble Sort ---
void bubbleSort(int arr[], int n) {
    int i, j, temp;
    int swapped;
    for (i = 0; i < n - 1; i++) {
        swapped = 0;
        for (j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = 1;
            }
        }
        if (swapped == 0) break;
    }
}

// --- Selection Sort ---
void selectionSort(int arr[], int n) {
    int i, j, min_idx, temp;
    for (i = 0; i < n - 1; i++) {
        min_idx = i;
        for (j = i + 1; j < n; j++) {
            if (arr[j] < arr[min_idx])
                min_idx = j;
        }
        if (min_idx != i) {
            temp = arr[min_idx];
            arr[min_idx] = arr[i];
            arr[i] = temp;
        }
    }
}

// --- Insertion Sort ---
void insertionSort(int arr[], int n) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

// --- Merge Sort ---
void merge_internal(int arr[], int l, int m, int r) {
    int i, j, k;
    int n1 = m - l + 1;
    int n2 = r - m;
    int *L = (int*)malloc(n1 * sizeof(int));
    int *R = (int*)malloc(n2 * sizeof(int));
    for (i = 0; i < n1; i++) L[i] = arr[l + i];
    for (j = 0; j < n2; j++) R[j] = arr[m + 1 + j];
    i = 0; j = 0; k = l;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) { arr[k] = L[i]; i++; }
        else { arr[k] = R[j]; j++; }
        k++;
    }
    while (i < n1) { arr[k] = L[i]; i++; k++; }
    while (j < n2) { arr[k] = R[j]; j++; k++; }
    free(L); free(R);
}

void mergeSort_internal(int arr[], int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;
        mergeSort_internal(arr, l, m);
        mergeSort_internal(arr, m + 1, r);
        merge_internal(arr, l, m, r);
    }
}

void mergeSort(int arr[], int n) {
    mergeSort_internal(arr, 0, n - 1);
}

// --- Quick Sort ---
int partition_internal(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

void quickSort_internal(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition_internal(arr, low, high);
        quickSort_internal(arr, low, pi - 1);
        quickSort_internal(arr, pi + 1, high);
    }
}

void quickSort(int arr[], int n) {
    quickSort_internal(arr, 0, n - 1);
}

// --- Randomized Quick Sort ---
int randomPartition_internal(int arr[], int low, int high) {
    int random = low + rand() % (high - low + 1);
    swap(&arr[random], &arr[high]);
    return partition_internal(arr, low, high);
}

void randomizedQuickSort_internal(int arr[], int low, int high) {
    if (low < high) {
        int pi = randomPartition_internal(arr, low, high);
        randomizedQuickSort_internal(arr, low, pi - 1);
        randomizedQuickSort_internal(arr, pi + 1, high);
    }
}

void randomizedQuickSort(int arr[], int n) {
    randomizedQuickSort_internal(arr, 0, n - 1);
}

// --- Heap Sort ---
void heapify_internal(int arr[], int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;
    if (left < n && arr[left] > arr[largest]) largest = left;
    if (right < n && arr[right] > arr[largest]) largest = right;
    if (largest != i) {
        swap(&arr[i], &arr[largest]);
        heapify_internal(arr, n, largest);
    }
}

void heapSort(int arr[], int n) {
    for (int i = n / 2 - 1; i >= 0; i--) heapify_internal(arr, n, i);
    for (int i = n - 1; i > 0; i--) {
        swap(&arr[0], &arr[i]);
        heapify_internal(arr, i, 0);
    }
}

// --- Shell Sort ---
void shellSort(int arr[], int n) {
    for (int gap = n / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < n; i++) {
            int temp = arr[i];
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                arr[j] = arr[j - gap];
            }
            arr[j] = temp;
        }
    }
}

// --- Counting Sort ---
void countingSort(int arr[], int n) {
    if (n <= 0) return;
    int max = arr[0], min = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > max) max = arr[i];
        if (arr[i] < min) min = arr[i];
    }
    int range = max - min + 1;
    int *count = (int*)calloc(range, sizeof(int));
    int *output = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) count[arr[i] - min]++;
    for (int i = 1; i < range; i++) count[i] += count[i - 1];
    for (int i = n - 1; i >= 0; i--) {
        output[count[arr[i] - min] - 1] = arr[i];
        count[arr[i] - min]--;
    }
    for (int i = 0; i < n; i++) arr[i] = output[i];
    free(count); free(output);
}

// --- Radix Sort ---
void countingSortForRadix_internal(int arr[], int n, int exp) {
    int *output = (int*)malloc(n * sizeof(int));
    int i, count[10] = {0};
    for (i = 0; i < n; i++) count[(arr[i] / exp) % 10]++;
    for (i = 1; i < 10; i++) count[i] += count[i - 1];
    for (i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }
    for (i = 0; i < n; i++) arr[i] = output[i];
    free(output);
}

void radixSort(int arr[], int n) {
    if (n <= 0) return;
    int max = arr[0];
    for (int i = 1; i < n; i++) if (arr[i] > max) max = arr[i];
    for (int exp = 1; max / exp > 0; exp *= 10)
        countingSortForRadix_internal(arr, n, exp);
}

// --- Benchmarking Logic ---

typedef void (*SortFunc)(int[], int);

typedef struct {
    char name[30];
    SortFunc func;
    bool is_slow; // O(n^2)
} Algorithm;

int main() {
    srand(42); // Deterministic seed for reproducibility

    Algorithm algos[] = {
        {"Bubble Sort", bubbleSort, true},
        {"Selection Sort", selectionSort, true},
        {"Insertion Sort", insertionSort, true},
        {"Merge Sort", mergeSort, false},
        {"Quick Sort", quickSort, false},
        {"Randomized Quick Sort", randomizedQuickSort, false},
        {"Heap Sort", heapSort, false},
        {"Shell Sort", shellSort, false},
        {"Counting Sort", countingSort, false},
        {"Radix Sort", radixSort, false}
    };
    int num_algos = sizeof(algos) / sizeof(algos[0]);

    int sizes[] = {100, 500, 1000, 2000, 5000, 10000, 20000, 50000, 100000, 200000, 500000, 1000000};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);

    FILE *fp = fopen("benchmark_results.csv", "w");
    if (!fp) {
        perror("Failed to open file");
        return 1;
    }
    fprintf(fp, "Algorithm,Size,Time\n");

    for (int i = 0; i < num_algos; i++) {
        printf("Benchmarking %s...\n", algos[i].name);
        for (int j = 0; j < num_sizes; j++) {
            int n = sizes[j];

            // Skip large sizes for O(n^2) algorithms to avoid excessive wait
            if (algos[i].is_slow && n > 100000) continue;

            // Allocate and generate random array
            int *arr = (int*)malloc(n * sizeof(int));
            for (int k = 0; k < n; k++) arr[k] = rand() % 1000000; // Limit range for Counting Sort

            // Validate correctness (once per algorithm at small size)
            if (j == 0) {
                int *temp_arr = (int*)malloc(n * sizeof(int));
                memcpy(temp_arr, arr, n * sizeof(int));
                algos[i].func(temp_arr, n);
                if (!is_sorted(temp_arr, n)) {
                    fprintf(stderr, "Error: %s failed validation!\n", algos[i].name);
                    free(temp_arr);
                    free(arr);
                    continue;
                }
                free(temp_arr);
            }

            double start = get_time_seconds();
            algos[i].func(arr, n);
            double end = get_time_seconds();
            double duration = end - start;

            fprintf(fp, "%s,%d,%f\n", algos[i].name, n, duration);
            fflush(fp);

            free(arr);

            // If it took too long (> 10s), skip remaining sizes for this slow algorithm
            if (algos[i].is_slow && duration > 10.0) {
                printf("  %s took too long (%.2fs), skipping larger sizes.\n", algos[i].name, duration);
                break;
            }
        }
    }

    fclose(fp);
    printf("Benchmarking complete. Results saved to benchmark_results.csv\n");

    return 0;
}
